---
title: Using the package
output: html_document
---

We provide within the package example datasets for both the single biopsy and the multi-regions/single-cell cases. We now load such example datasets.

```{r}
library("ASCETIC")
data(dataset_example_single_samples)
data(ccf_dataset_example_single_samples)
data(dataset_example_phylogenies)
data(models_phylogenies)
```

First, we outline the inference steps for the single biopsy NGS data scenario, where only one sample per patient is available and the temporal patters among genes are estimated considering cancer cell fractions data. To test this, we consider an example dataset comprising 10 hepatocellular carcinoma sequencing samples from Nguyen, Bastien, et al., 2022. We recall that such (reduced and partial) dataset was provided only for computational testing and should not be used for biological analysis/conclusions.

We now perform the basic ASCETIC inference.

```{r}
set.seed(12345)
res_example_single_samples = perform.ascetic.ccf.dataset( 
                                dataset = dataset_example_single_samples, 
                                ccf.dataset = ccf_dataset_example_single_samples, 
                                regularization = "aic", 
                                command = "hc", 
                                restarts = 0 )
```

The output of this analysis is a a list of 4 elements: 1) dataset, input dataset; 2) ccf dataset, input ccf dataset; 3) poset, partially order set among mutations estimated by ASCETIC from the agony ranking and 4) inference, inferred ASCETIC evolutionary model for each selected regularization.

To improve the statistical stability of the framework expecially in the common situation of noisy data, ASCETIC implements a re-sampling algorithm which builds on the function just described. This procedure is more computational intensive compared to the basic one, but should be preferred on real data. It requires an additional dataset with variant allele frequencies information and can be executed as follow.

```{r}
set.seed(12345)
data(vaf_dataset_example_single_samples)
res_example_single_samples_resampling = perform.ascetic.ccf.dataset.resampling( 
                                            dataset = dataset_example_single_samples, 
                                            ccf.dataset = ccf_dataset_example_single_samples, 
                                            vaf.dataset = vaf_dataset_example_single_samples, 
                                            nsampling = 5, 
                                            regularization = "aic", 
                                            command = "hc", 
                                            restarts = 0 )
```

We refer to the manual for a detailed description of each parameter.

We now describe the execution of the ASCETIC framework when high-resolution cancer genomics data are available, such as multi-regions data or single-cell data. In this case, phylogenetic reconstructions for each patient are available and are given as input to ASCETIC. We consider an other example dataset consisting of 10 myeloid malignancies samples sequenced with Tapestri from Miles, Linde A., et al., 2020. We recall that also in this case, such (reduced and partial) dataset was provided only for computational testing and should not be used for biological analysis/conclusions.

We now perform the basic ASCETIC inference for this second scenario.

```{r}
set.seed(12345)
res_example_phylogenies_dataset = perform.ascetic.phylogenies.dataset( 
                                    dataset = dataset_example_phylogenies, 
                                    models = models_phylogenies, 
                                    regularization = "aic", 
                                    command = "hc", 
                                    restarts = 0 )
```

The output of this analysis is a a list of 4 elements: 1) dataset, input dataset; 2) models, input phylogenetic models; 3) poset, partially order set among mutations estimated by ASCETIC from the agony ranking and 4) inference, inferred ASCETIC evolutionary model for each selected regularization.

Also in this case we provide a method to improve the statistical stability of the framework expecially in the situation of noisy data. ASCETIC implements a bootstrap procedure which builds on the function just described. This procedure should be preferred on real data. It can be executed as follow.

```{r}
set.seed(12345)
res_example_phylogenies_dataset_bootstrap = perform.ascetic.phylogenies.dataset.bootstrap( 
                                                dataset = dataset_example_phylogenies, 
                                                models = models_phylogenies, 
                                                nsampling = 5, 
                                                regularization = "aic", 
                                                command = "hc", 
                                                restarts = 0 )
```

We refer to the manual for a detailed description of each parameter and to the ASCETIC manuscript for details on the method.
