%\VignetteEngine{knitr::knitr}

\documentclass[a4paper,9pt]{article}

<<style-knitr, eval=TRUE, echo=FALSE, results="asis">>=
BiocStyle::latex()
@

%\VignetteIndexEntry{ASCETIC}

\usepackage[utf8]{inputenc}
\usepackage{graphicx}
\usepackage{placeins}
\usepackage{url}
\usepackage{tcolorbox}
\usepackage{authblk}

\begin{document}

\title{An R package for the inference of statistically significant temporal patterns involving alternations in driver genes from cancer genomics data}

\author[1]{Luca De Sano}
\author[2]{Daniele Ramazzotti}

\affil[1]{Department of Informatics, Systems and Communication, Università degli Studi Milano Bicocca Milano, Italy.}
\affil[2]{Department Medicine and Surgery, Università degli Studi Milano Bicocca Milano, Italy.}

\date{\today}
\maketitle

\begin{tcolorbox}{\bf Overview.}
Cancer development is a stochastic process involving large populations of cells. Random genetic and epigenetic alterations commonly occurring in any cell can occasionally be beneficial to neoplastic ones, thus defining clones characterized by a functional selective advantage. During clonal evolution, certain clones can be positively selected for increased proliferation and survival ability, outgrowing competing cells and this can eventually lead to invasion and metastasis. Throughout such a multi-step stochastic process, cancer cells can acquire over time a set of biological capabilities that sometimes are referred to as hallmarks. Not all variants are involved in their acquisition, but only a relative small subset of them – i.e., the drivers –, while most mutations present in the cancer clones do not increase their fitness – i.e., the passengers.
ASCETIC (Agony-baSed Cancer EvoluTion InferenCe) is a novel framework for the inference of a set of statistically significant temporal patterns involving alternations in driver genes from cancer genomics data.
\vspace{1.0cm}

{\em In this vignette, we give an overview of the package by presenting some of its main functions.}

\vspace{1.0cm}

\renewcommand{\arraystretch}{1.5}

\end{tcolorbox}

<<include=FALSE>>=
library(knitr)
opts_chunk$set(
concordance = TRUE,
background = "#f3f3ff"
)
@

\newpage

\tableofcontents

\section{Changelog} 

\begin{itemize}
\item[1.0.0] Package released in October 2022.
\end{itemize}

\section{Using the ASCETIC R package}

ASCETIC is an R package for the inference of statistically significant temporal patterns involving alternations in driver genes. The framework can be applied to data at different resolutions, from classical NGS genomics data of cancer samples, to single-cell data.

In its basic implementation, ASCETIC requires two main inputs: (i) a binary matrix where rows are patients (i.e., samples) and columns are mutations. Each cell of the matrix is 1 if the related mutation was observed in the sample; 0 otherwise. (2) Information of temporal patters across the considered driver genes. Such information can be determined either from classical NGS genomics data considering a single biopsy per patient, in terms of cancer cell fractions reported for each gene or from data at either resolutions such as multi-region or single-cell data where multiple samples per patients are sequenced; in this latter case, ASCETIC takes as input a phylogenetic tree per patient.

We provide within the package example datasets for both the single biopsy and the multi-regions/single-cell cases. We now load such example datasets. 

<<req>>=
library("ASCETIC")
data(dataset_example_single_samples)
data(ccf_dataset_example_single_samples)
data(dataset_example_phylogenies)
data(models_phylogenies)
@

First, we outline the inference steps for the single biopsy NGS GS genomics data scenario, where only one sample per patient is available and the temporal patters among genes are estimated considering cancer cell fractions data. To test this, we consider an example dataset comprising 10 hepatocellular carcinoma sequencing samples from Nguyen, Bastien, et al., 2022. We recall that such (reduced and partial) dataset was provided only for computational testing and should not be used for biological analysis/conclusions.

We now perform the basic ASCETIC inference.

<<step1,eval=TRUE>>=
set.seed(12345)
res_example_single_samples = perform.ascetic.ccf.dataset( dataset = dataset_example_single_samples, 
                                                          ccf.dataset = ccf_dataset_example_single_samples, 
                                                          regularization = "aic", 
                                                          command = "hc", 
                                                          restarts = 0 )
@

The output of this analysis is a a list of 4 elements: 1) dataset, input dataset; 2) ccf dataset, input ccf dataset; 3) poset, partially order set among mutations estimated by ASCETIC from the agony ranking and 4) inference, inferred ASCETIC evolutionary model for each selected regularization.

To improve the statistical stability of the framework expecially in the common situation of noisy data, ASCETIC implements a re-sampling algorithm which builds on the function just described. This procedure is more computational intensive compared to the basic one, but should be preferred on real data. It requires an additional dataset with variant allele frequencies information and can be executed as follow.

<<step2,eval=TRUE>>=
set.seed(12345)
data(vaf_dataset_example_single_samples)
res_example_single_samples_resampling = perform.ascetic.ccf.dataset.resampling( dataset = dataset_example_single_samples, 
                                                                                ccf.dataset = ccf_dataset_example_single_samples, 
                                                                                vaf.dataset = vaf_dataset_example_single_samples, 
                                                                                nsampling = 5, 
                                                                                regularization = "aic", 
                                                                                command = "hc", 
                                                                                restarts = 0 )
@

We refer to the manual for a detailed description of each parameter.

We now describe the execution of the ASCETIC framework when high-resolution cancer genomics data are available, such as multi-regions data or single-cell data. In this case, phylogenetic reconstructions for each patient are available and are given as input to ASCETIC. We consider an other example dataset consisting of 10 myeloid malignancies samples sequenced with Tapestri from Miles, Linde A., et al., 2020. We recall that also in this case, such (reduced and partial) dataset was provided only for computational testing and should not be used for biological analysis/conclusions.

We now perform the basic ASCETIC inference for this second scenario.

<<step3,eval=TRUE>>=
set.seed(12345)
res_example_phylogenies_dataset = perform.ascetic.phylogenies.dataset( dataset = dataset_example_phylogenies, 
                                                                       models = models_phylogenies, 
                                                                       regularization = "aic", 
                                                                       command = "hc", 
                                                                       restarts = 0 )
@

The output of this analysis is a a list of 4 elements: 1) dataset, input dataset; 2) models, input phylogenetic models; 3) poset, partially order set among mutations estimated by ASCETIC from the agony ranking and 4) inference, inferred ASCETIC evolutionary model for each selected regularization.

Also in this case we provide a method to improve the statistical stability of the framework expecially in the situation of noisy data. ASCETIC implements a bootstrap procedure which builds on the function just described. This procedure should be preferred on real data. It can be executed as follow.

<<step4,eval=TRUE>>=
set.seed(12345)
res_example_phylogenies_dataset_bootstrap = perform.ascetic.phylogenies.dataset.bootstrap( dataset = dataset_example_phylogenies, 
                                                                                           models = models_phylogenies, 
                                                                                           nsampling = 5, 
                                                                                           regularization = "aic", 
                                                                                           command = "hc", 
                                                                                           restarts = 0 )
@

<<sessioninfo,results='asis',echo=FALSE>>=
toLatex(sessionInfo())
@

\end{document}
